Tools/clinic/clinic.py

```diff
@@ -910,54 +940,98 @@ def deprecate_positional_use(
+                amount = f"{first_pos+1} "
+            message = f"Parameter {pstr} will become a keyword-only parameter"
         else:
             condition = f"nargs > {first_pos} && nargs <= {last_pos+1}"
             if first_pos:
-                preamble = (
-                    f"Passing more than {first_pos} positional "
-                    f"argument{'s' if first_pos != 1 else ''} to "
-                )
-            depr_message = preamble + (
-                f"{func.fulldisplayname}() is deprecated. Parameters {pstr} will "
-                f"become keyword-only parameters in Python {major}.{minor}."
-            )
-
+                amount = f"more than {first_pos} "
+                if first_pos == 1:
+                    what = "argument"
+            message = f"Parameters {pstr} will become keyword-only parameters"
+        message = (
+            f"Passing {amount}positional {what} to {func.fulldisplayname}() is deprecated. "
+            f"{message} in Python {major}.{minor}."
+        )
         # Append deprecation warning to docstring.
-        lines = textwrap.wrap(f"Note: {depr_message}")
-        docstring = "\n".join(lines)
+        docstring = textwrap.fill(f"Note: {message}")
         func.docstring += f"\n\n{docstring}\n"
+        # Format and return the code block.
+        code = self.DEPRECATION_WARNING_PROTOTYPE.format(
+            condition=condition,
+            errcheck="",
+            message=wrapped_c_string_literal(message, width=64,
+                                             subsequent_indent=20),
+        )
+        return normalize_snippet(code, indent=4)
+
+    def deprecate_keyword_use(
+            self,
+            func: Function,
+            params: dict[int, Parameter],
+    ) -> str:
+        assert len(params) > 0
+        names = [repr(p.name) for p in params.values()]
+        first_pos, first_param = next(iter(params.items()))
+        last_pos, last_param = next(reversed(params.items()))
+
+        # Pretty-print list of names.
+        pstr = pprint_words(names)
 
+        # For now, assume there's only one deprecation level.
+        assert first_param.deprecated_keyword == last_param.deprecated_keyword
+        thenceforth = first_param.deprecated_keyword
+        assert thenceforth is not None
+        major, minor = thenceforth
+
+        # Format the deprecation message.
+        containscheck = ""
+        conditions = []
+        for i, p in params.items():
+            if p.is_optional():
+                #conditions.append(f"nargs < {i+1} && {argname_fmt % i}")
+                if func.kind.new_or_init:
+                    conditions.append(f"nargs < {i+1} && kwargs && PyDict_Contains(kwargs, &_Py_ID({p.name}))")
+                    containscheck = "PyDict_Contains"
+                else:
+                    conditions.append(f"nargs < {i+1} && kwnames && PySequence_Contains(kwnames, &_Py_ID({p.name}))")
+                    containscheck = "PySequence_Contains"
```

Comment:
Regarding `_Py_ID`:

- See commit 6f6a4e6cc5cd76af4a53ffbb62b686142646ac9a
- gh-95860
- gh-90928
- gh-95907

Reply:
cc. @ericsnowcurrently

Reply:
What's the concern with using `_Py_ID()` here?

Reply:
Currently, there is no concern; I just wanted to point out that even though Argument Clinic does not currently care about the Limited API, we _may_ want to add such functionality in the future; IMO, it should be possible for Argument Clinic to generate code that does not depend on internal APIs, for example for stdlib extension modules such as sqlite3.

Many core devs seem to be in favour of not using internal APIs in stdlib extension modules, but I'm not sure if there is a consensus.